pkg/alert/alert.go

```diff
@@ -0,0 +1,327 @@
+// the given timestamp.
+func (a *Alert) ResolvedAt(ts time.Time) bool {
+	if a.EndsAt.IsZero() {
+		return false
+	}
+	return !a.EndsAt.After(ts)
+}
+
+// Queue is a queue of alert notifications waiting to be sent. The queue is consumed in batches
+// and entries are dropped at the front if it runs full.
+type Queue struct {
+	logger       log.Logger
+	maxBatchSize int
+	labels       labels.Labels
+
+	mtx   sync.Mutex
+	queue []*Alert
+	morec chan struct{}
+
+	pushed  prometheus.Counter
+	popped  prometheus.Counter
+	dropped prometheus.Counter
+}
+
+// NewQueue returns a new queue. The given label set is attached to all alerts pushed to the queue.
+func NewQueue(logger log.Logger, reg prometheus.Registerer, capacity, maxBatchSize int, lset labels.Labels) *Queue {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+	q := &Queue{
+		logger:       logger,
+		queue:        make([]*Alert, 0, capacity),
+		morec:        make(chan struct{}, 1),
+		maxBatchSize: maxBatchSize,
+		labels:       lset,
+
+		dropped: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_alert_queue_alerts_dropped_total",
+			Help: "Total number of alerts that were dropped from the queue.",
+		}),
+		pushed: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_alert_queue_alerts_pushed_total",
+			Help: "Total number of alerts pushed to the queue.",
+		}),
+		popped: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "thanos_alert_queue_alerts_popped_total",
+			Help: "Total number of alerts popped from the queue.",
+		}),
+	}
+	capMetric := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "thanos_alert_queue_capacity",
+		Help: "Capacity of the alert queue.",
+	}, func() float64 {
+		return float64(q.Cap())
+	})
+	lenMetric := prometheus.NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "thanos_alert_queue_length",
+		Help: "Length of the alert queue.",
+	}, func() float64 {
+		return float64(q.Cap())
+	})
+	if reg != nil {
+		reg.MustRegister(q.pushed, q.popped, q.dropped, lenMetric, capMetric)
+	}
+	return q
+}
+
+// Len returns the current length of the queue.
+func (q *Queue) Len() int {
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+	return len(q.queue)
+}
+
+// Cap returns the fixed capacity of the queue.
+func (q *Queue) Cap() int {
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+	return cap(q.queue)
+}
+
+// Pop takes a batch of alerts from the front of the queue. The batch size is limited
+// according to the queues maxBatchSize limit.
+// It blocks until elements are available or a termination signal is send on termc.
+func (q *Queue) Pop(termc <-chan struct{}) []*Alert {
+	select {
+	case <-termc:
+		return nil
+	case <-q.morec:
+	}
+
+	q.mtx.Lock()
+	defer q.mtx.Unlock()
+
+	as := make([]*Alert, q.maxBatchSize)
+	n := copy(as, q.queue)
+	q.queue = q.queue[n:]
+
+	q.popped.Add(float64(n))
+
+	return as[:n]
+}
+
+// Push adds a list of alerts to the queue.
+func (q *Queue) Push(alerts ...*Alert) {
```

Comment:
hm, I would avoid doing variadic slice here. Just pass slice. it does not make sense to do `Push()`. Do you remember our bug with `prometheus.Register()`? (: